realm-transformer/src/main/kotlin/io/realm/transformer/IncrementalBuild.kt

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2018 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.transformer
+
+import com.android.SdkConstants
+import com.android.build.api.transform.Status
+import com.android.build.api.transform.TransformInput
+import com.android.build.api.transform.TransformOutputProvider
+import io.realm.annotations.RealmClass
+import javassist.CtClass
+import javassist.NotFoundException
+import org.gradle.api.Project
+import java.io.File
+import java.util.jar.JarFile
+
+class IncrementalBuild(project: Project, outputProvider: TransformOutputProvider, transform: RealmTransformer)
+    : TransformBuildTemplate(project, outputProvider, transform) {
+
+    override fun prepareOutputClasses(inputs: MutableCollection<TransformInput>) {
+        this.inputs = inputs;
+        findClassNames(inputs, outputClassNames, outputReferencedClassNames) // Output files
+        logger.debug("Incremental build. Files being processed: ${outputClassNames.size}.")
+        logger.debug("Incremental files: ${outputClassNames.joinToString(",")}")
+    }
+
+    /**
+     * Go through the transform input in order to find all files we need to transform.
+     *
+     * @param inputs set of input files
+     * @param directoryFiles the set of files in directories getting compiled. These are candidates for the transformer.
+     * @param referencedFiles the set of files that are possible referenced but never transformed (required by JavaAssist).
+     * @param isIncremental `true` if build is incremental.
+     */
+    override fun findClassNames(inputs: Collection<TransformInput>,
+                                directoryFiles: MutableSet<String>,
+                                referencedFiles: MutableSet<String>) {
+        inputs.forEach {
+            // Files in directories are files we most likely want to transform
+             it.directoryInputs.forEach {
+                val dirPath: String = it.file.absolutePath
+                it.changedFiles.entries.forEach {
+                    if (it.value == Status.NOTCHANGED || it.value == Status.REMOVED) {
+                        return
```

Comment:
Why `return`? isn't `continue` enough, we want to keep iterating the other files in the dir

Reply:
good catch. This actually exits the method. In kotlin you cannot call continue in a lambda but need to do `return@forEach`